File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class ListNode{
    int val;
    ListNode next;

    public ListNode(int val, ListNode next){
        this.val = val;
        this.next = next;
    }

 }

class BSTIterator {
    ListNode head;
    ListNode dummy;
    public BSTIterator(TreeNode root) {
        head = new ListNode(-1,null);
        dummy = head;
        createLinkedList(root);        
    }
    
    public int next() {
        head = head.next;
        return head.val;
    }
    
    public boolean hasNext() {
        if(head.next != null) return true;

        return false;
    }

    private void createLinkedList(TreeNode root){
        //base
        if(root == null){
            return;
        }

        createLinkedList(root.left);
        ListNode node = new ListNode(root.val, null);
        dummy.next = node;
        // System.out.println(dummy.val + "before");
        dummy = dummy.next;
        // System.out.println(dummy.val + "after");
        createLinkedList(root.right);
    }
}
//Time Complexity: O(n)
// Space Complexity: O(n)
/**
 * Your BSTIterator object will be instantiated and called as such:
 * BSTIterator obj = new BSTIterator(root);
 * int param_1 = obj.next();
 * boolean param_2 = obj.hasNext();
 */

File: 0173-binary-search-tree-iterator/0173-binary-search-tree-iterator.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class ListNode:
    # val = None
    # next = None
    def __init__(self, val, next):
        self.val = val
        self.next = next

class BSTIterator:
    dummy = None
    head = None
    def __init__(self, root: Optional[TreeNode]):
        self.printroot = root
        self.head = ListNode(-1, None)
        self.dummy = self.head
        self.createLinkedList(root)
        # print(self.dummy.next)      
    
        # print(self.head.next.val)

    def next(self) -> int:
        self.head = self.head.next
        return self.head.val

    def hasNext(self) -> bool:
        if self.head.next:
            return True
        
        return False
    
    def createLinkedList(self, root):
        
        #base
        if root is None:
            return
        # print(root)
        self.createLinkedList(root.left)
        # print(root.val)
        node = ListNode(root.val, None)
        self.dummy.next = node
        # print( self.dummy.val, "before")
        self.dummy = self.dummy.next
        # print(self.dummy.val, "new dummy node")
        self.createLinkedList(root.right)



# Your BSTIterator object will be instantiated and called as such:
# obj = BSTIterator(root)
# param_1 = obj.next()
# param_2 = obj.hasNext()